<?xml version="1.0" ?>
<launch>

    <!-- Load the robot description parameter. -->
    <!-- This is used by the robot hardware interface and by the Gazebo plugin. -->
    <param name="robot_description"
           command="$(find xacro)/xacro.py '$(find trikey_controlit)/models/trikey_controlit/xacro/trikey_controlit.xacro'"/>

    <!-- Specify the servo clock type. -->
    <!-- <rosparam param="servo_clock_type">controlit_servo_clock/ServoClockROS</rosparam> -->
    <rosparam param="servo_clock_type">controlit_servo_clock/ServoClockChrono</rosparam>

    <!-- Specify the servo clock frequency. -->
    <rosparam param="servo_frequency">1000</rosparam>

    <!-- Specify the robot interface type. -->
    <!-- <rosparam param="robot_interface_type">controlit_robot_interface/RobotInterfaceSM</rosparam> -->
    <!-- <rosparam param="robot_interface_type">controlit_robot_interface/RobotInterfaceROSTopic</rosparam> -->
    <rosparam param="robot_interface_type">controlit_robot_interface/RobotInterfaceUDP</rosparam>

    <!-- Specify the odometry topic -->
    <!-- <rosparam param="odometry_topic">/gazebo/root_link_odom</rosparam> -->
    <rosparam param="odometry_topic">/trikey/gazebo/root_link_odom</rosparam>

    <!-- Two types of controllers are currently supported: Torque and Impedance -->
    <rosparam param="whole_body_controller_type">controlit_wbc/WBOSC</rosparam>

    <rosparam param="parameter_binding_factories">["controlit_binding_factory/BindingFactoryROS", "controlit_binding_factory/BindingFactorySM"]</rosparam>

    <rosparam param="state_delay_ms">0</rosparam>
    <rosparam param="command_delay_ms">0</rosparam>

    <!-- Specify the log level and the log prefix fields -->
    <rosparam param="log_level">DEBUG</rosparam>
    <rosparam param="log_fields">["package", "file", "line", "function"]</rosparam>

    <param name="RobotInterfaceUDP/address" type="str" value="127.0.0.1" />
    <param name="RobotInterfaceUDP/cmdPort" type="int" value="51124" />
    <param name="RobotInterfaceUDP/statePort" type="int" value="51125" />

    <!-- The following enable/disable multi-threading in ControlIt! -->
    <!--
    <rosparam param="USE_SINGLE_THREADED_CONTROL_MODEL">true</rosparam>
    <rosparam param="USE_SINGLE_THREADED_TASK_UPDATER">true</rosparam>
    <rosparam param="USE_SINGLE_THREADED_SENSOR_UPDATER">true</rosparam>
    -->

    <!--
       Here are some other parameters that should be set:

       controlit_max_effort_command
       controlit_gravity_vector
       controlit_ramp_up_time' does not exist
       controlit_coupled_joint_groups
       controlit_gravity_compensation_mask
       controlit_effective_gain_pass_through_mask
       controlit_model_delta_cutoff
       controlit_model_blend_rate
    -->

</launch>
